consensus/src/txn_manager.rs

```diff
@@ -50,28 +52,37 @@ impl TxnManager for MempoolProxy {
                 });
             }
         }
-        let (callback, callback_rcv) = oneshot::channel();
-        let req = ConsensusRequest::GetBlockRequest(max_size, exclude_txns, callback);
-        // send to shared mempool
-        self.consensus_to_mempool_sender
-            .clone()
-            .try_send(req)
-            .map_err(anyhow::Error::from)?;
-        // wait for response
-        match monitor!(
-            "pull_txn",
-            timeout(Duration::from_secs(1), callback_rcv).await
-        ) {
-            Err(_) => {
-                Err(format_err!("[consensus] did not receive GetBlockResponse on time").into())
+        let no_pending_txns = exclude_txns.is_empty();
+        for _ in 0..50 {
+            let (callback, callback_rcv) = oneshot::channel();
+            let req = ConsensusRequest::GetBlockRequest(max_size, exclude_txns.clone(), callback);
+            // send to shared mempool
+            self.consensus_to_mempool_sender
+                .clone()
+                .try_send(req)
+                .map_err(anyhow::Error::from)?;
+            // wait for response
+            let txns = match monitor!(
+                "pull_txn",
+                timeout(Duration::from_secs(1), callback_rcv).await
+            ) {
+                Err(_) => Err(MempoolError::from(anyhow::anyhow!(
+                    "[consensus] did not receive GetBlockResponse on time"
+                ))),
+                Ok(resp) => match resp.map_err(anyhow::Error::from)?? {
+                    ConsensusResponse::GetBlockResponse(txns) => Ok(txns),
+                    _ => Err(MempoolError::from(anyhow::anyhow!(
+                        "[consensus] did not receive expected GetBlockResponse"
+                    ))),
+                },
+            }?;
+            if txns.is_empty() && no_pending_txns {
+                delay_for(Duration::from_millis(NO_TXN_DELAY)).await;
```

Comment:
500 ms delay, how will this translate in practice on testnet or a local run for example?
can we make this configurable? and if we do, could we make the configuration call into two different functions?
that way we can disable it if it becomes problematic on mainnet
also could a 500 ms delay ever translate into a timeout? should instead should we aim for a 1/10 or 1/4 time of the expected timeout (assuming that is relatively universal)
final thought... should we do 5 * 100 how much cpu are we burning doing this 50 times versus 5?

Reply:
Why didn't we make this configurable btw?